main/src/library/MutList.flix

```diff
@@ -56,27 +58,35 @@ instance Foldable[MutList] {
 
 namespace MutList {
 
+    ///
+    /// Constant which stores the minimum capacity of a MutList.
+    ///
+    def minCapacity(): Int32 = 8
+
     ///
     /// Returns a new empty mutable list with a default capacity.
     ///
-    pub def new(): MutList[a] & Impure = MutList(ref [], ref 0)
+    pub def new(): MutList[a] & Impure = MutList(ref [default; minCapacity()], ref 0)
 
     ///
     /// Returns a mutable list of all integers between `b` (inclusive) and `e` (exclusive).
     ///
     /// Returns an empty mutable list if `b >= e`.
     ///
-    pub def range(b: Int32, e: Int32): MutList[Int32] & Impure = MutList(ref Array.range(b, e), ref (e - b))
+    pub def range(b: Int32, e: Int32): MutList[Int32] & Impure =
+        let minCap = minCapacity();
+        if (e - b < minCap)
+            let f = i -> if (b + i < e) b + i else default;
+            MutList(ref Array.init(f, minCap), ref (e - b))
+        else
+            MutList(ref Array.range(b, e), ref (e - b))
```

Comment:
@jaschdoc @mlutze But it is of course also confusing to say that there is a notion of "minCapacity" and that they it might not always be available. But indeed when a list is created and then pushed to, it no longer has minCapacity free slots! Maybe the name should be `defaultCapacity` or `initialCapacity`?

Reply:
If I'm reading this correctly, then we indeed get an array of 1-10, not 1-18.
My understanding is that we only have extra space if the length is less than the `minCapacity`. E.g. `range(1, 4)` has an array of size 8.

So in my view this makes sense: the minimum capacity of the backing array is 8.

Reply:
I'm fairly certain, this works as @mlutze describes. I think renaming the `minCapacity` to something else won't solve the problem of ambiguity. We probably just need to write some documentation explaining that the array is never shorter than length 8.

Reply:
But it is certainly not the minimum capacity, since the space available can be less than that. I think `defaultCapacity` might be more descriptive.

Whats the point of the if branch? Why can't this just always be ` MutList(ref Array.range(b, e), ref (e - b))`?

Reply:
> Whats the point of the if branch? Why can't this just always be MutList(ref Array.range(b, e), ref (e - b))?

If the range between `b` to `e` is less than the `minCapacity` / `defaultCapacity` then the array size would be less than that which would be an illegal state for MutList now that we have defined that the size of the array in a MutList can never go below 8.

Reply:
I guess there's a bit of a question about the definition of "capacity"? To me it refers to the amount of space _total_, even if part or all of that is occupied. For example, a room with a capacity of 10 people still has a capacity of 10, whether there are 0, 7, or 10 people inside.

So for me, `minCapacity` seems like the most appropriate.

Reply:
I think originally the semantics was what @mlutze describes. Here is what you said @magnus-madsen in #2009:
> Intuitively there is a number MIN_SIZE and this is the initial size and a shrink can never go below that.

My interpretation of that was that the array can never go below `minCapacity`. Maybe we should rename it to `minSize` instead?